sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala

```diff
@@ -90,13 +90,12 @@ private[sql] abstract class SparkStrategies extends QueryPlanner[SparkPlan] {
            left.statistics.sizeInBytes <= sqlContext.conf.autoBroadcastJoinThreshold =>
           makeBroadcastHashJoin(leftKeys, rightKeys, left, right, condition, joins.BuildLeft)
 
-      // If the sort merge join option is set, we want to use sort merge join prior to hashjoin
-      // for now let's support inner join first, then add outer join
-      case ExtractEquiJoinKeys(Inner, leftKeys, rightKeys, condition, left, right)
+      // If the sort merge join option is set, we want to use sort merge join prior to hashjoin.
+      // And for outer join, we can not put conditions outside of the join
+      case ExtractEquiJoinKeys(joinType, leftKeys, rightKeys, condition, left, right)
         if sqlContext.conf.sortMergeJoinEnabled =>
-        val mergeJoin =
-          joins.SortMergeJoin(leftKeys, rightKeys, planLater(left), planLater(right))
-        condition.map(Filter(_, mergeJoin)).getOrElse(mergeJoin) :: Nil
+        joins.SortMergeJoin(
+          leftKeys, rightKeys, joinType, planLater(left), planLater(right), condition) :: Nil
 
```

Comment:
Shall we move the code to a new `Strategy `(like SortMergeJoin) instead of mix in   `Hashjoin`?